### PR TITLE
Import job: remote (SSH) archive destinations (import/process split PR 2, Task 2.7)

### PR DESCRIPTION
**Pull request by jss367**
## What

Completes **Task 2.7** of the import-job plan (`docs/plans/2026-07-04-import-job-plan.md`), which #1107 deferred: an **opt-in SSH remote archive destination** for the new import job (`run_import_job`). The local/SMB import path is byte-for-byte unchanged; this adds a second transport.

This is the "support SSH as well as local/SMB" piece — SSH is the more robust transport for pushing a shoot to the NAS over the network (the SMB stalls that motivated the import/process split are an SMB artifact), while the archive stays browsable at the mount path and hash-verify remains an opt-in knob.

## How

- **Transfer:** rsync card → `remote_path/subpath` over SSH; catalog rows land at the browsable `mount_path/subpath` (`resolve_remote_archive` mapping). No local staging tree — `move._run_rsync_streamed` gains an opt-in `src_specs` file-list mode.
- **Verification is genuinely card → NAS.** `move.remote_verify_files` runs `rsync -an --checksum <card files> <nas target>` and fails only the specific card file the NAS didn't confirm. Gated behind `verify_by_hash`; without it, rows keep NULL `hash_status`/`hash_checked_at` and `safe_to_format` honestly reports `False` with reason `"enable verify_by_hash for remote verification"`. A mount-vs-NAS check would be tautological (same physical storage), so it is deliberately avoided — the "safe to format" pill must mean exactly what it says (CORE_PHILOSOPHY: no black boxes).
- **Same-basename collision parity** with the local path: colliding files land under numeric-suffixed names on the NAS (each transferred individually to an explicit filename) instead of silently clobbering.
- **Route:** `POST /api/jobs/import-photos` accepts `remote_target_id` + `remote_subpath`, mirroring the pipeline route's guards (mutual exclusivity with `destination`, unknown-target 404, GNU-rsync availability refusal); the transport is snapshotted at enqueue.

### Known scope boundary
Cross-folder duplicate-twin linking is simplified for the remote path (only the current batch's dest folder is linked, not arbitrary twin folders elsewhere in the archive). Documented inline. Honesty is preserved (remote-without-verify is already `safe=False`; with-verify, duplicates are byte-confirmed).

## Tests

Full plan suite: **1847 passed, 2 skipped** (pre-existing platform skips). `ruff check vireo/ tests/` clean. New coverage: card→NAS verify targets the card paths (not the mount), verify-failure fails the specific file and blocks safe-to-format, no-verify leaves hashes NULL + unsafe, and same-basename collision lands distinct rows.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added remote SSH/rsync “archive destination” support for photo imports, including per-batch transfers and workspace linking for verified duplicates.

* **Bug Fixes**
  * Improved remote import request validation (mutual exclusivity, safe subpath rules, non-empty target requirements) and clearer errors for missing/unsupported rsync.
  * Strengthened remote integrity behavior: verification stamps hashes when enabled and rolls back/fails only the specific mismatched landings.

* **Tests**
  * Expanded end-to-end coverage for remote rsync routing, batching, verification modes, collisions, and edge-case failure handling.

### CODE REVIEW

File: vireo/import_job.py [hunk 1]
```diff
@@ -379,6 +392,542 @@ def _linkable_twin_dirs(rows, under_destination):
...
+                            _fail(rel, source_file,
+                                  f"duplicate check failed: {e}")
+                            continue
+                    accept = False
+                    for twin in twin_rows:
+                        twin_path = os.path.join(
+                            twin["folder_path"], twin["filename"],
+                        )
+                        try:
+                            twin_hash = compute_file_hash(twin_path)
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py]: Addressed in `a079722b`. The remote duplicate gate now filters twins under any import source root before accepting: `import_job.py` line ~744, `if _path_under_any_source(twin_path): continue` inside the twin-accept loop, mirroring the local path. A source-card twin can no longer back a skip.

File: vireo/import_job.py [hunk 2]
```diff
@@ -379,6 +482,566 @@ def _linkable_twin_dirs(rows, under_destination):
...
+                ]
+                renamed = [
+                    (sf, bn) for sf, bn in to_transfer
+                    if bn != sf.name
+                ]
+
+                def _do_rsync(src_specs, target, dest_is_dir, extra_args):
+                    try:
+                        rc, stderr, timed_out = move_mod._run_rsync_streamed(
+                            None, target, [], len(src_specs), None,
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py:1176]: Addressed in `b07aebef`. The per-batch rsync now passes `--ignore-existing` (see the `extra_args` list, `import_job.py` line ~844), so the receiver skips files already present on the NAS and a raced second writer cannot clobber the first landing. On the verify path the subsequent `--checksum` step then fails that specific file honestly.

File: vireo/import_job.py [hunk 3]
```diff
@@ -379,6 +482,638 @@ def _linkable_twin_dirs(rows, under_destination):
...
+
+        dest_folder = (
+            os.path.normpath(os.path.join(destination, rel))
+            if rel != "." else destination
+        )
+        ssh_dest = (
+            posixpath.join(ssh_base, *rel.split("/")) if rel != "."
+            else ssh_base
+        )
+        os.makedirs(dest_folder, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py:794]: Addressed in `3b6b7066`. After `os.makedirs(dest_folder)` the remote path now runs `UPDATE folders SET status = 'ok' WHERE path = ? AND status = 'missing'` (`import_job.py` line ~688), promoting a stale `missing` row so the imported photos are visible in workspace queries. Preserves `partial` (a real needs-rescan signal), mirroring the local path.

File: vireo/import_job.py [hunk 4]
```diff
@@ -379,6 +482,638 @@ def _linkable_twin_dirs(rows, under_destination):
...
+                        db.update_photo_hash_check(
+                            row["id"], "ok", commit=False,
+                        )
+                else:
+                    copied -= 1
+                    if params.verify_by_hash:
+                        verified -= 1
+                    _counts(rel)["copied"] -= 1
+                    _fail(rel, dest_path,
+                          "not cataloged after scan (no photo row)")
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py:1527]: Addressed in `3b6b7066`. The catalog-row presence check now does a companion lookup before failing: when `row is None`, it queries `SELECT p.id ... WHERE f.path = ? AND p.companion_path = ?` (`import_job.py` line ~1173) and accepts the paired JPEG whose sibling RAW adopted it as `companion_path`, instead of failing it as `not cataloged after scan`. Mirrors the local path.

File: vireo/import_job.py [hunk 5]
```diff
@@ -379,6 +482,712 @@ def _linkable_twin_dirs(rows, under_destination):
...
+                    # skip as an intra-batch duplicate; otherwise advance.
+                    if claimed_basenames[candidate] == src_hash:
+                        skipped_duplicate += 1
+                        dup_skipped += 1
+                        _counts(rel)["skipped_duplicate"] += 1
+                        adopted = True
+                        break
+                    counter += 1
+                    continue
+                if os.path.exists(cand_mount):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py:1052]: Addressed in `c3fe0a89`. There is now a batch-level dest-under-source guard before `os.makedirs`: `if _path_under_any_source(dest_folder):` fails the whole batch (`import_job.py` line ~647), so when the mount base is an ancestor of a source and the template maps back into it, no source-backed `cand_mount` is ever hashed as a duplicate. Mirrors the local path.

File: vireo/import_job.py [hunk 6]
```diff
@@ -379,6 +482,771 @@ def _linkable_twin_dirs(rows, under_destination):
...
+        # cataloged+linked (a duplicate-only batch would otherwise leave the
+        # mount folder invisible). scan() over the mount is the same call the
+        # local path makes; the mount is locally walkable.
+        if landed or dup_skipped:
+            landed_paths = {entry[0] for entry in landed}
+            try:
+                scan(
+                    destination, db,
+                    restrict_dirs=[dest_folder],
+                    restrict_files=(landed_paths or None),
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py]: Fixed in `7461db67`. The collision loop's crash-recovery adopt branch now tracks the pre-existing mount path in an `adopted_paths` set (`import_job.py` line ~890) and the batch scan uses `restrict_files=landed_paths | adopted_paths` (line ~1065), so an adopted-but-uncataloged file in a mixed batch is scanned and gets a photo row instead of staying invisible while `copied + skipped_duplicate == discovered`. Its photo id is also added to `imported_photo_ids` for chaining. Covered by a new mixed-batch test.

File: vireo/import_job.py [hunk 7]
```diff
@@ -379,6 +482,771 @@ def _linkable_twin_dirs(rows, under_destination):
...
+                        except OSError:
+                            continue
+                        if twin_hash is not None and twin_hash == src_hash:
+                            accept = True
+                            break
+                    if accept:
+                        skipped_duplicate += 1
+                        dup_skipped += 1
+                        _counts(rel)["skipped_duplicate"] += 1
+                        continue
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py:976]: Fixed in `7461db67` (real visibility bug, not just a scope boundary). When an accepted duplicate's cataloged twin lives under the archive but in a different folder than this run's template output, the batch scan (which only walks `dest_folder`) never linked it. The remote path now collects those twin dirs via `_linkable_twin_dirs` scoped to the destination and runs a follow-up `scan(restrict_dirs=…)` link pass (line ~827 + the dup-link scan), and a link failure flips `dup_link_failed` so `safe_to_format`/`ok` go False — mirroring the local path's gate. Covered by a new test: a verified duplicate-only skip whose twin is in a different archive layout is now linked into the workspace.

File: vireo/import_job.py [hunk 8]
```diff
@@ -379,6 +482,771 @@ def _linkable_twin_dirs(rows, under_destination):
...
+        })
+    safe_to_format = (
+        not cancelled
+        and failed == 0
+        and not discovery_errors
+        and not partial_scope
+        and not remote_unverified
+        and (copied + skipped_duplicate) == discovered
+    )
+    result = {
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_job.py:1870]: Fixed in `7461db67`. `_run_remote_import_job` now tracks freshly cataloged mount rows, RAW primaries that adopted a landed JPEG companion, and adopted crash-recovery duplicates in `imported_photo_ids`, and returns `"photo_ids": sorted(imported_photo_ids)` in the result (`import_job.py` line ~1427) — the same key/shape the local path uses. `_chain_after_import` (`result.get("photo_ids")`) now enqueues the follow-up process job for remote imports with new photos. Duplicate-only imports still return an empty list (chaining falls into "no new photos" as designed). Covered by a new test.

**Comment by jss367** [comment]
🤖 The PR agent was fired for `address-review` (author: `jss367`), but the review body was empty and the seven inline review comments are all self-acknowledgments confirming your own fixes to earlier Codex bot findings (e.g. "Addressed in `a079722b`", "Fixed in `7461db67`"). There's no reviewer-requested change I can honestly translate into code, so I'm stopping without pushing a commit.

For your awareness, two Codex P2 findings from the 01:15:55 UTC review on commit `7461db67` don't yet have a reply on this PR:

- `vireo/import_job.py` line 1169 — "Reject non-empty rows with missing scan hashes" (verify-by-hash path stamps `ok` when `scan()` leaves `file_hash` NULL).
- `vireo/import_job.py` line 1216 — "Validate companion bytes without remote verification" (paired-JPEG rows accepted in no-verify mode without hashing the mount JPEG).

If either should be addressed here, reply with a review comment saying so (or comment `/claude-fix` with a directive) and I'll take another pass.

---
_Generated by [Claude Code](https://claude.ai/code)_
